### ISSUE TITLE
Redisson codec: java.lang.ClassNotFoundException: org.springframework.util.MultiValueMapAdapter

### ISSUE THREAD

**Issue by willzgw** [redisson/redisson#5725]
I have a Java service based on SpringBoot 2.2.5 and redisson-spring-data-22 3.15.4,
and I am trying to upgrade the SpringBoot to 2.7.18 to Support JDK 21, when I gray-release the new service(Old and new versions will co-exist), the old version reported the following error:
```
ERROR CommandDecoder - [bTxId: , bSpanId: , bMode: ] Unable to decode data. channel: [id: 0x7d1672fe, L:/10.169.66.210:20614 - R:/10.169.75
.15:6379], reply: ReplayingDecoderByteBuf(ridx=2034, widx=2034), command: (EVAL), params: [local value = redis.call('hget', KEYS[1], ARGV[2]); if value == false then return nil; end; lo
cal t,..., 5, gw_PrestoQueryRecord, redisson__timeout__set:{gw_PrestoQueryRecord}, redisson__idle__set:{gw_PrestoQueryRecord}, redisson__map_cache__last_access__set:{gw_PrestoQueryRecor
d}, {gw_PrestoQueryRecord}:redisson_options, 1711353891504, PooledUnsafeDirectByteBuf(ridx: 0, widx: 57, cap: 256)] (CommandDecoder.java:180)
java.io.IOException: java.lang.ClassNotFoundException: org.springframework.util.MultiValueMapAdapter
        at org.redisson.codec.MarshallingCodec.lambda$new$0(MarshallingCodec.java:148)
        at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:358)
        at org.redisson.client.handler.CommandDecoder.decodeCommand(CommandDecoder.java:177)
        at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:116)
        at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:101)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:498)
        at io.netty.handler.codec.ReplayingDecoder.callDecode(ReplayingDecoder.java:366)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.ClassNotFoundException: org.springframework.util.MultiValueMapAdapter
        at org.springframework.boot.web.embedded.tomcat.TomcatEmbeddedWebappClassLoader.loadClass(TomcatEmbeddedWebappClassLoader.java:72)
        at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1188)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:468)
        at org.jboss.marshalling.AbstractClassResolver.loadClass(AbstractClassResolver.java:129)
        at org.jboss.marshalling.AbstractClassResolver.resolveClass(AbstractClassResolver.java:110)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadClassDescriptor(RiverUnmarshaller.java:1048)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1381)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:298)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:246)
        at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1879)
        at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1793)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1421)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:298)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:246)
        at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1879)
        at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1793)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1421)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:298)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:231)
        at org.jboss.marshalling.AbstractObjectInput.readObject(AbstractObjectInput.java:41)
        at org.redisson.codec.MarshallingCodec.lambda$new$0(MarshallingCodec.java:146)
        ... 23 common frames omitted
```

And the new version reported:
```
ERROR CommandDecoder - Unable to decode data. channel: [id: 0xc64ac7ec, L:/10.169.67.100:37398 - R:/10.169.75.15:6379], reply: ReplayingDecoderByteBuf(ridx=2115, widx=2115), command: (EVAL), params: [local value = redis.call('hget', KEYS[1], ARGV[2]); if value == false then return nil; end; local t,..., 5, gw_PrestoQueryRecord, redisson__timeout__set:{gw_PrestoQueryRecord}, redisson__idle__set:{gw_PrestoQueryRecord}, redisson__map_cache__last_access__set:{gw_PrestoQueryRecord}, {gw_PrestoQueryRecord}:redisson_options, 1711381047995, PooledUnsafeDirectByteBuf(ridx: 0, widx: 57, cap: 256)] (CommandDecoder.java:180)
java.io.IOException: java.lang.ClassNotFoundException: org.springframework.util.CollectionUtils$MultiValueMapAdapter
	at org.redisson.codec.MarshallingCodec.lambda$new$0(MarshallingCodec.java:148)
	at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:358)
	at org.redisson.client.handler.CommandDecoder.decodeCommand(CommandDecoder.java:177)
	at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:116)
	at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:101)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)
	at io.netty.handler.codec.ReplayingDecoder.callDecode(ReplayingDecoder.java:366)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.ClassNotFoundException: org.springframework.util.CollectionUtils$MultiValueMapAdapter
	at org.springframework.boot.web.embedded.tomcat.TomcatEmbeddedWebappClassLoader.loadClass(TomcatEmbeddedWebappClassLoader.java:72)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1141)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:468)
	at org.jboss.marshalling.AbstractClassResolver.loadClass(AbstractClassResolver.java:129)
	at org.jboss.marshalling.AbstractClassResolver.resolveClass(AbstractClassResolver.java:110)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadClassDescriptor(RiverUnmarshaller.java:1048)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1381)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:298)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:246)
	at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1879)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1793)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1421)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:298)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:246)
	at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1879)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1793)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1421)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:298)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:231)
	at org.jboss.marshalling.AbstractObjectInput.readObject(AbstractObjectInput.java:41)
	at org.redisson.codec.MarshallingCodec.lambda$new$0(MarshallingCodec.java:146)
	... 23 common frames omitted
```
Not sure if this is due to compatibility between SpringBoot and redisson-spring-data-22?
Looking forward to your reply, thanks in advance.

**Comment by mrniko** [comment]
I think the old version misses "org.springframework.util.MultiValueMapAdapter"

**Comment by willzgw** [comment]
Hi @mrniko 
Thanks for you reply.
I just found that, this is the problem with our service design:
When we read and write to Redis, we used the default JAVA serialization instead of JSON serialization, which writes the relevant class path into the serialization, because the class path changed between SpringBoot 2.2 and SpringBoot 2.7, which makes it impossible for newer versions to find each other. class written to it.
Below is an example of Java serialization:
```
\x00\x00\x00\x00\x00\x00\x00\x00\xac\x0c\x00\x00\x00\x00\x00\x00\x04\x04\t><com.shopee.di.presto.gateway.model.gateway.PrestoQueryRecord\x13-pdF\xe5N\x1c\x00\x00\x00\x13>\bcatalogs\x16\x00>\ncreateTime\x16\x00>\bhttpBody\x16\x00>\x0bhttpHeaders\x16\x00>\nhttpMethod\x16\x00>\x02id\x16\x00>\x03idc\x16\x00>\bpriority#\x00>\rqueryCostType\x16\x00>\x0cqueryResults\x16\x00>\tqueryType\x16\x00>\x05queue\x16\x00>\rredirectedIdc\x16\x00>\x06status\x16\x00>\x0esubmitDirectly \x00>\x0csubmitStatus\x16\x00>\tuniqueKey\x16\x00>\nupdateTime\x16\x00>\buserName\x16\x00\x16\x048>\x11java.util.HashSet\xbaD\x85\x95\x96\xb8\xb74\x00\x00\x00\x00\x162\x0c\x00\x00\x00\x10?@\x00\x00\x00\x00\x00\x005L\x00\x00\x01\x8eyv\x98\xd6>m------------------\xe6\x89\x8b\xe6\x9c\xba\xe5\x8f\xb7\xe6\xae\xb5\xe8\x81\x9a\xe9\x9b\x86----------------\ndrop table if exists tmp_henry_reg_ana_risk_feature_phone_pre_clu_lm\x04\t>$org.springframework.http.HttpHeaders\x88\xf2\xd9\xf2w\xc9T\xa4\x00\x00\x00\x01>\aheaders\x16\x00\x16\x04\t>=org.springframework.util.CollectionUtils$MultiValueMapAdapteraBF\xa5\xeb3\xcc\xae\x00\x00\x00\x01>\x03map\x16\x00\x16\x04\t>1org.springframework.util.LinkedCaseInsensitiveMap\x18\\\x92\x0c\xdd\x83\xa3\x0e\x00\x00\x00\x03>\x13caseInsensitiveKeys\x16\x00>\x06locale\x16\x00>\ttargetMap\x16\x00\x16\x048>\x11java.util.HashMap\x05\a\xda\xc1\xc3\x16`\xd1\x00\x00\x00\x02>\nloadFactor&\x00>\tthreshold#\x00\x16?@\x00\x00\x00\x00\x00\x182\b\x00\x00\x00 \x00\x00\x00\x12>\x10x-presto-catalog9\xff>\x0econtent-length9\xff>\x0fx-presto-source>\x0fX-Presto-Source>\x11x-forwarded-proto9\xff>\rx-presto-user>\rX-Presto-User>\x14presto-fake-query-id>\x14Presto-Fake-Query-ID>\x1cx-presto-client-capabilities9\xff>\tforwarded9\xff>\x17x-presto-transaction-id9\xff>\x0fx-presto-schema9\xff>\x10x-presto-session>\x10X-Presto-Session>\x04host>\x04Host>\x0ccontent-type9\xff>\x12x-presto-time-zone9\xff>\x14x-presto-client-tags>\x14X-Presto-Client-Tags>\x11x-presto-language9\xff>\x0faccept-encoding9\xff>\nuser-agent9\xff5\x048>\x10java.util.Locale~\xf8\x11`\x9c0\xf9\xec\x00\x00\x00\x06>\acountry\x16\x00>\nextensions\x16\x00>\bhashcode#\x00>\blanguage\x16\x00>\x06script\x16\x00>\avariant\x16\x00\x16==\xff\xff\xff\xff>\x02en==5\x04\t>3org.springframework.util.LinkedCaseInsensitiveMap$1\xa5\xe5e\xe0\x80\xe9[\xd2\x00\x00\x00\x01>\x06this$0\x16\x00\t>\x17java.util.LinkedHashMap4\xc0N\\\x10l\xc0\xfb\x00\x00\x00\x01>\x0baccessOrder \x00;\xfc?@\x00\x00\x00\x00\x00\x182\b\x00\x00\x00 \x00\x00\x00\x129\xe4S\x01[>-host=presto-gw.prod.data-infra.shopee.io:84439\xdcS\x01[>\x05https9\xd7S\x01[>\x031219\xd4S\x01[>\x04hive9\xdeS\x01[>\x11dev_security_data9\xe2S\x01[>\rAsia/Shanghai9\xe3S\x01[>\x05en-US9\xd7S\x01[>\x04NONE9\xd3S\x01[>\x18PATH,PARAMETRIC_DATETIME9\xdfS\x01[>\x1cPresto JDBC Driver/338-dirty9\xd7S\x01[>\x19text/plain; charset=utf-89\xdaS\x01[>\x04gzip9\xcaS\x01[>DSG_regantifraud-adhoc__gateway__87716ba7_1ad0_4d19_8fe0_c32a9074473b9\xc3S\x01[>L(25)-(regantifraud-adhoc)-(xingting.jiang)-(scheduler)-(xingting.jiang)-(SG)9\xcc\x04\t>6com.google.common.collect.ImmutableList$SerializedForm\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01>\belements\x16\x00\x16B\x01\x16>/hive.insert_existing_partitions_behavior=APPEND9\xc9S\x01[>\x1210.169.74.211:84439\xbdS\x01[>\x0exingting.jiang9\xc9S\x01[?\x01jenv:dev,proj:security_data,tic:dev.security_data.studio_2223406_20240326_DAY_1.0,pic:xingting.jiang,pri:25,open_auth:true,submit_time:1711434799318,fake_query_id:SG_regantifraud-adhoc__gateway__87716ba7_1ad0_4d19_8fe0_c32a9074473b,dispatch_info:10.169.74.211:8443#[default]#stt#strictly,query_cost_type:META,submit_directly:true,original_client_user:di_scheduler5\x009\xae\x04\x0c\x00\x00\x00#org.springframework.http.HttpMethod\x00\x00\x00\x04POST\x01>\x03stt\x00\x00\x00\x19>\x04META?\x02D{\"id\":\"20240326_063319_07905_97iym\",\"infoUri\":\"https://10.169.74.211:8443/ui/query.html?20240326_063319_07905_97iym\",\"nextUri\":\"https://10.169.74.211:8443/v1/statement/queued/20240326_063319_07905_97iym/y14f7d78e9c0c9fee56f052e0dd392522287eeaae/1\",\"stats\":{\"state\":\"QUEUED\",\"queued\":true,\"scheduled\":false,\"nodes\":0,\"totalSplits\":0,\"queuedSplits\":0,\"runningSplits\":0,\"completedSplits\":0,\"cpuTimeMillis\":0,\"wallTimeMillis\":0,\"queuedTimeMillis\":0,\"elapsedTimeMillis\":0,\"processedRows\":0,\"processedBytes\":0,\"physicalInputBytes\":0,\"peakMemoryBytes\":0,\"spilledBytes\":0},\"warnings\":[]}\n>\x0fDATA_DEFINITION>\x15SG_regantifraud-adhoc\x01K\x00\x00\x00\xc8\x01K\x00\x00\x00\x029\xecL\x00\x00\x01\x8eyv\x98\xf19\xf7
```
As you see, the class path `org.springframework.util.CollectionUtils$MultiValueMapAdapter` in included, which can not be found in SpringBoot 2.7.
